system/doc/reference_manual/expressions.xml

```diff
@@ -836,6 +836,28 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
       depending on the size of the float because otherwise comparison of large
       floats and integers would lose their transitivity.</p>
 
+    <p>The term equivalence operators, <c>=:=</c> and <c>=/=</c>, return
+      whether two terms are indistinguishable. While the other operators
+      consider the same <em>numbers</em> equal even when their types differ
+      (<c>1 == 1.0</c> is true), the term equivalence operators return whether
+      there exists any function that can tell their arguments apart.</p>
```

Comment:
`∀A∀B ¬(A =/= B) ≡ (A =:= B) ≡ ∀f (f(A) =:= f(B)))`, what's the problem? :grin:

> "... whether there exists **no** function...", 

The operators are mirrors of each other, so they both say the same thing, and I'm not sure the extra negation helps. :thinking: 

> Maybe just stick with the conventional notion of "... are exactly the same...", 

Hacker News got all up in arms because they misunderstood what we meant by "exactly equal to," so I'm not sure that'll help much either. :sweat_smile:

Reply:
I definitely prefer term-equivalence operators as a name. I took a stab at writing this section. Feel free use any part of it or throw it all out. The idea is that we would describe the general rules, as they are currently written, provide examples, and then discuss comparing numbers in their own section.

---

The rules above apply to all comparison operators. The distinction from `==` and `/=` to `=:=` and `=/=` arises when comparing integers and floats. Let's explore this case in more detail.

## Comparing numbers

Similar to `+` and other arithmetic operators, Erlang provides a series of equality operators to compare numbers, regardless if they are integers or floats:

```erlang
1> 1 == 1.0.
true
2> 1 /= 1.0.
false
3> 10 < 5.0.
false
4> 10 >= 5.
true
```

When comparing an integer to a float, the number with the lesser precision is converted into the type of the other. A float is more precise than an integer until all significant figures of the float are to the left of the decimal point. This happens when the float is larger/smaller than `+/-9007199254740992.0`. The conversion strategy is changed depending on the size of the float because otherwise comparison of large floats and integers would lose their transitivity.

Additionally, floating-point numbers have an interesting property where zero is either negative (-0.0) or positive (+0.0), although in practice they compare equal:

```erlang
1> +0.0 == -0.0
true
```

However, in some cases, you may want to distinguish between integers and floats, or between positive and negative zero, as they have different in-memory representations (and in some mathematical domains they may also exhibit different properties). Erlang provides two _term equivalence operators_ (`=:=` and `=/=`) for this exact purpose:

```erlang
1> 1 =:= 1.0.
false
2> +0.0 =:= -0.0
false
```

Finally, note pattern matching relies on term equivalence. Therefore, the following expressions won't match either:

```erlang
1> 1 = 1.0.
** exception error: no match of right hand side value 1.0
2> -0.0 = +0.0.
```

To avoid ambiguity, Erlang will warn if 0.0 is used inside a pattern. You must instead match on `+0.0` or `-0.0`. If you want to match on both positive and negative zero floating points, you can write:

```erlang
is_floating_zero(Num) when is_float(Num), Num == 0.0 -> true;
is_floating_zero(_Num) -> false.
```

In other words, the comparison operators will behave in accordance to the other arithemetic operators in the language, and adhere to the IEEE 754 specification. The term equivalence operators allow developers to explicitly compare their underlying representations.

> Note: Prior to OTP 27, the term equivalence operators had a bug where they considered <c>0.0</c> and <c>-0.0</c> to be the same term.

Reply:
FWIW, Elixir calls those operators "strictly equal". It may make more sense to think some comparisons are more strict than others (rather than more exact).

Reply:
Thanks, I like that take on things. I'll try to incorporate it when I'm back from vacation mid-to-late September.